changes needed in cobra to support the coreos/pflags

joshi4 opened the pull request:
Because cobra was written with pflags in mind, changing pflags makes it necessary to change cobra. The strip flags function works on the POSIX style of flags. I found that instead of rewriting that function to support the much wider range of flags accepted now, it would be much easier and quicker to add a reformatArgs() function which would take in arguments in the original goflags style and convert them to the style that pflags expects. 

Major changes to review: reformatArgs(), getAllBooleanFlags()

joshi4 committed:
changed imports

joshi4 committed:
using joshi4/goflags library now

joshi4 committed:
removed f.Shorthand field

joshi4 committed:
wrote a function to reformat arguments into posix style , which can then be easily interpretted by cobra

joshi4 committed:
removed extra debug print statements

joshi4 committed:
correctly reformat arguments after a flag terminator now

joshi4 committed:
reformatArgs is more idiomatic now

joshi4 committed:
Merge branch 'modified_pflags'

joshi4 committed:
changing import

joshi4 committed:
added in shorthands to DebugFlags()

philips commented on command.go:
```diff
@@ -375,6 +378,68 @@ func (c *Command) errorMsgFromParse() string {
 }
 
+//Takes the command line arguments in all the various forms that
+//etcdctl supports and converts the flags to POSIX style flags
```
gofmt here

philips commented on command.go:
```diff
@@ -397,6 +465,9 @@ func (c *Command) Execute() (err error) {
 		args = c.args
 	}
 
+	//reformat the arguments here:
```
useless comment.

philips commented on command.go:
```diff
@@ -405,7 +476,12 @@ func (c *Command) Execute() (err error) {
 			c.Usage()
 		}
 	} else {
+		// fmt.Println("args to findAndExecute are : ", args)
```
please don't checkin uncommented code.

philips commented on command.go:
```diff
@@ -405,7 +476,12 @@ func (c *Command) Execute() (err error) {
+		if err != nil {
+
+			// fmt.Println("finished findAndExecute with error = ", err.Error())
+		}
```
Hrm, this should be deleted too.

philips commented on command.go:
```diff
@@ -713,6 +790,50 @@ func (c *Command) ParseFlags(args []string) (err error) {
+	return baseSlice
+}
+
+// call only on the root.
```
somewhat useless comment here too

philips commented on command.go:
```diff
@@ -713,6 +790,50 @@ func (c *Command) ParseFlags(args []string) (err error) {
+
+// call only on the root.
+func (c *Command) getAllBooleanFlags() (error, map[string]struct{}) {
+
```
newline cleanup

philips commented on command.go:
```diff
@@ -713,6 +790,50 @@ func (c *Command) ParseFlags(args []string) (err error) {
+
+	}
+
+	// adding persistent flags now:
```
ueless comment

philips commented on command.go:
```diff
@@ -280,6 +281,7 @@ func (c *Command) Find(arrs []string) (*Command, []string, error) {
 	innerfind = func(c *Command, args []string) (*Command, []string) {
 		if len(args) > 0 && c.HasSubCommands() {
 			argsWOflags := stripFlags(args)
+			// fmt.Println("Arguments without flags are ", argsWOflags)
```
please don't checkin commented out code.

joshi4 committed:
removed useless comments, commented out code and other suggested fixes

changed test case, so expected behaviour is to not parse a short flag correctly if there is no space between the flag and it's value

changed  test case to reflect updated error message, now that long flags can be preceded by a single dash

default library does not support specifying a short flag's value without a space

test case no longer applicable

removing more extraneous comments

changed behaviour of reformatArgs to fail explicitly incase of badly specified flags, current behaviour was to simply omit the badly formatted flags from the reformated args list

PR fixes, adapting test suite and explicit error reporting during reformatArgs

joshi4 committed:
Merge branch 'PRfix'

joshi4 committed:
forgot to remove pesky print stmt
